test/test_model.py

```diff
@@ -2657,8 +2759,13 @@ def test_juju_log(self):
                          [['juju-log', '--log-level', 'BAR', '--', 'foo']])
 
     def test_valid_metrics(self):
+        if typing.TYPE_CHECKING:
+            _caselist = typing.List[typing.Tuple[
+                typing.Mapping[str, typing.Union[int, float]],
+                typing.Mapping[str, str],
+                typing.List[typing.List[str]]]]
```

Comment:
Similar here: clearer and simpler to just write out the test cases. Again, not your issue, but might as well clean it up and avoid the messy type annotation.

I don't mind case-based tests (what Go calls "table-driven tests"), but I think they should only be used when there are really a lot of cases and the case-based approach is clearer.